### ISSUE TITLE
Wrong(?) message "Argument incompatible with supertype" when method receive a instance of the same parametrized class

### ISSUE THREAD

**Issue by mfcabrera** [python/mypy#8877]
Note: if you are reporting a wrong signature of a function or a class in
the standard library, then the typeshed tracker is better suited
for this report: https://github.com/python/typeshed/issues

Please provide more information to help us understand the issue:

* Are you reporting a bug, or opening a feature request?
I am reported what I believe is a bug  but it could be I am using the annotation wrong (or maybe a workaround is documented). 

* Please insert below the code you are checking with mypy,
  or a mock-up repro if the source is private. We would appreciate
  if you try to simplify your case to a minimal repro.

```python
from typing import TypeVar, Generic
from abc import abstractmethod, ABC

K = TypeVar('K')


class Metric(ABC, Generic[K]):

    @abstractmethod
    def sum(self, other: "Metric[K]") -> "Metric[K]":
        pass


class MyFloatMetric(Metric[float]):
    def sum(self, other: "MyFloatMetric") -> "MyFloatMetric":
        return MyFloatMetric()
````

* What is the actual behavior/output?
```console
sample2.py:15:5: error: Argument 1 of "sum" is incompatible with supertype "Metric"; supertype defines the argument type as "Metric[float]"  [override]
```

* What is the behavior/output you expect?

As MyFloatMetric is a actually a Metric[float] mypy should not mark this as an error.
If I replace thee `MyFloatMetric` for `Metric[float]` as the first parameter it works. However doing this in a hierachy of classes where on different levels the Generic types are fixed might be complicated. 

* What are the versions of mypy and Python you are using?
  Do you see the same issue after installing mypy from Git master?

mypy 0.770
python 3.8.1 

On master I get a similar error message:

```console
sample2.py:15:5: error: Argument 1 of "sum" is incompatible with supertype "Metric"; supertype defines the argument type as "Metric[float]"  [override]
sample2.py:15:5: note: This violates the Liskov substitution principle
sample2.py:15:5: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides
```

* What are the mypy flags you are using? (For example --strict-optional)
From `tox.ini` I have:

```ini
[mypy]

# Global settings
disallow_untyped_defs    = False
show_column_numbers      = True
show_error_codes         = True
strict_optional          = True
warn_no_return           = True
warn_redundant_casts     = True
# activate later
warn_unused_ignores      = False
ignore_missing_imports   = True

# Enable these over time
check_untyped_defs       = False
disallow_incomplete_defs = False
no_implicit_optional     = False
warn_return_any          = False
warn_unreachable         = False
```

**Comment by hynekcer** [comment]
A solution that was possible even for mypy 0.770 is with a bound TypeVar:

``` python
K = TypeVar('K')
TSameMetric = TypeVar("TSameMetric", bound="Metric")

class Metric(ABC, Generic[K]):
    @abstractmethod
    def sum(self: TSameMetric, other: TSameMetric) -> TSameMetric:
        pass
```
It is important that the type of `other` parameter was restricted to the same type still in the abstract parent class. (Otherwise subclasses would violate the Liskov substitution principle and would be correctly rejected.)

Your `MyFloatMetric` is the same. I add some examples how it works with subclasses of MyFloatMetric and other Metric. I think that inheritance works now how you expected. 
``` python
class MySubFloatMetric(MyFloatMetric):
    def sum(self, other: "MyFloatMetric") -> "MySubFloatMetric":
        return MySubFloatMetric()

class YourFloatMetric(Metric[float]):
    def sum(self, other: "YourFloatMetric") -> "YourFloatMetric":
        return YourFloatMetric()

class ComplexMetric(Metric[complex]):
    def sum(self, other: "ComplexMetric") -> "ComplexMetric":
        return ComplexMetric()

a = MyFloatMetric()
a.sum(a)

sa = MySubFloatMetric()
sa.sum(sa)
a.sum(sa)
sa.sum(a)

y = YourFloatMetric()
y.sum(y)
a.sum(y)  # correctly fails

c = ComplexMetric()
c.sum(c)
a.sum(c)  # correctly fails
```
Python 3.11(currently rc1) implements a special type [Self](https://docs.python.org/3.11/library/typing.html#typing.Self) that should allow a much clearer declaration with the same functionality,
``` python
from typing import Self

class Metric(ABC, Generic[K]):
    @abstractmethod
    def sum(self: Self, other: Self) -> Self:
        pass
```
but currently is the support for`Self` actively developed in mypy and ready yet.

**Comment by erictraut** [comment]
I think mypy's behavior is correct here. It should generate an error here because parameters are implicitly contravariant, and `MyFloatMetric` is a subtype (not a supertype) of `Metric[float]`.
